servers/rendering/renderer_rd/shader_rd.cpp

```diff
@@ -191,9 +191,14 @@ void ShaderRD::_build_variant_code(StringBuilder &builder, uint32_t p_variant, c
 				for (const KeyValue<StringName, CharString> &E : p_version->code_sections) {
 					builder.append(String("#define ") + String(E.key) + "_CODE_USED\n");
 				}
-#if defined(MACOS_ENABLED) || defined(IOS_ENABLED)
-				builder.append("#define MOLTENVK_USED\n");
+#if (defined(MACOS_ENABLED) || defined(IOS_ENABLED))
+				if (RD::get_singleton()->get_device_capabilities().device_family == RDD::DEVICE_VULKAN) {
+					builder.append("#define MOLTENVK_USED\n");
+				}
+				// Image atomics are supported on Metal 3.1 but no support in MoltenVK or SPIRV-Cross yet.
+				builder.append("#define NO_IMAGE_ATOMICS\n");
```

Comment:
I don't know much about this but I wonder if this is related to #95448 ?

Reply:
Yes, this fixes issues with MoltenVK too 👍🏻 I made some adjustments in the shaders, so that we split MOLTENVK conditional defines with image atomics specifically. It was @reduz who originally put me on to the artefacts caused by image atomics.

Reply:
Is there any other use for `MOLTENVK_USED`, it was added specifically to work around missing image atomics and have no other purpose, so keeping only generic `NO_IMAGE_ATOMICS` should be sufficient.